runtime/near-wallet-contract/implementation/wallet-contract/src/tests/emulation.rs

```diff
@@ -0,0 +1,145 @@
+
+// The Wallet Contract should understand that transactions to other Wallet
+// Contract instances are base token transactions.
+#[tokio::test]
+async fn test_base_token_transfer() -> anyhow::Result<()> {
+    const TRANSFER_AMOUNT: NearToken = NearToken::from_near(2);
+
+    let TestContext { worker, wallet_contract, wallet_sk, wallet_contract_bytes, .. } =
+        TestContext::new().await?;
+
+    let (other_wallet, other_address) =
+        TestContext::deploy_wallet(&worker, &wallet_contract_bytes).await?;
+
+    let initial_wallet_balance = wallet_contract.inner.as_account().view_account().await?.balance;
+    let initial_other_balance = other_wallet.inner.as_account().view_account().await?.balance;
+
+    let transaction = aurora_engine_transactions::eip_2930::Transaction2930 {
+        nonce: 0.into(),
+        gas_price: 0.into(),
+        gas_limit: 0.into(),
+        to: Some(Address::new(other_address)),
+        value: Wei::new_u128(TRANSFER_AMOUNT.as_yoctonear() / u128::from(MAX_YOCTO_NEAR)),
+        data: b"A message for the recipient".to_vec(),
+        chain_id: CHAIN_ID,
+        access_list: Vec::new(),
+    };
+    let signed_transaction = crypto::sign_transaction(transaction, &wallet_sk);
+
+    let result =
+        wallet_contract.rlp_execute(other_wallet.inner.id().as_str(), &signed_transaction).await?;
+
+    assert!(result.success);
+
+    let final_wallet_balance = wallet_contract.inner.as_account().view_account().await?.balance;
+    let final_other_balance = other_wallet.inner.as_account().view_account().await?.balance;
+
+    // Receiver balance increases
+    assert_eq!(
+        final_other_balance.as_yoctonear(),
+        initial_other_balance.as_yoctonear() + TRANSFER_AMOUNT.as_yoctonear()
+    );
+    // Sender balance decreases (by a little more than the
+    // `TRANSFER_AMOUNT` due to gas spent to execute the transaction)
+    let diff = NearToken::from_yoctonear(
+        initial_wallet_balance.as_yoctonear()
+            - (final_wallet_balance.as_yoctonear() + TRANSFER_AMOUNT.as_yoctonear()),
+    );
+    assert!(diff < NearToken::from_millinear(2));
+
+    Ok(())
+}
+
+// The Wallet Contract should understand the ERC-20 standard and map
+// it to NEP-141 function calls.
+#[tokio::test]
+async fn test_erc20_emulation() -> anyhow::Result<()> {
+    const MINT_AMOUNT: NearToken = NearToken::from_near(100);
+    const TRANSFER_AMOUNT: NearToken = NearToken::from_near(32);
+
+    let TestContext {
+        worker,
+        wallet_contract,
+        wallet_sk,
+        wallet_address,
+        wallet_contract_bytes,
+        ..
+    } = TestContext::new().await?;
+
+    let token_contract = nep141::Nep141::deploy(&worker).await?;
+    token_contract.mint(wallet_contract.inner.id(), MINT_AMOUNT.as_yoctonear()).await?;
+
+    // Check balance
+    let transaction = aurora_engine_transactions::eip_2930::Transaction2930 {
+        nonce: 0.into(),
+        gas_price: 0.into(),
+        gas_limit: 0.into(),
```

Comment:
Yes, these parameters are currently not validated by the wallet contract. We could validate the `gas_limit` against the `prepaid_gas` host function in the Near runtime, but unfortunately there is no way to get the gas price from inside the runtime to validate against the transaction's gas price.

This does give relayers room to lie to users about the cost of transactions, but I don't think it is a critical vulnerability because the relayers do not profit from misrepresenting the gas price.